### PR TITLE
[torch] Add backward support for segment reduce (CPU only)

### PR DESCRIPTION
**Pull request by serhaty**
Summary:
This is to setup boiler plate code for backward and CPU implementation.

Next Steps in order:
- Add backward support for CUDA
- Add support for more aggregation types
- Benchmarking (for cuda mainly)/more testing/documentation
- Support for multi dimension

Test Plan:
Updated unit test to also check correctness of backward.

Wait for CI signal

Differential Revision: D27970340

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -8945,4 +8945,9 @@
 - func: segment_reduce(Tensor data, str reduce, *, Tensor? lengths=None, Tensor? indices=None, int axis=0, bool unsafe=False) -> Tensor
   variants: function
   dispatch:
-    CPU: _segment_reduce_cpu
+    CPU, CUDA: segment_reduce_kernel
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:8948]: This function looks pretty fresh. Consider adding an out= variant and making it structured, see https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 2]
```diff
@@ -60,6 +61,54 @@ Tensor _segment_reduce_cpu_kernel(
   return output;
 }
 
+Tensor _segment_reduce_cpu_backward_kernel(
+    const Tensor& grad,
+    const Tensor& output,
+    const Tensor& data,
+    const Tensor& lengths) {
+  auto grad_output = at::zeros({data.sizes()}, grad.options());
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:69]: This variable should be called `grad_input` (it is gradient with respect to operation input), `grad_output` or `grad` is the gradient you receive

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 3]
```diff
@@ -60,6 +61,54 @@ Tensor _segment_reduce_cpu_kernel(
...
+        auto* grad_output_data = grad_output.data_ptr<scalar_t>();
+        const auto* values_data = data_contig.data_ptr<scalar_t>();
+        int64_t k = 0;
+        for (int64_t i = 0; i < batch_size; ++i) {
+          // scalar_t reduction = std::numeric_limits<scalar_t>::lowest();
+          int64_t counter = 0;
+          for (int64_t j = 0; j < lengths_data[i]; ++j) {
+            if (at::_isnan(values_data[k]) ||
+                values_data[k] == output_data[i]) {
+              grad_output_data[k] = 1;
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:93]: `grad_output_data[k] = grad[i]`, and compare `grad_output_data[index] !=0` in the next loop. Also, as pointed above, the name should be `grad_input`.

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 4]
```diff
@@ -60,6 +61,54 @@ Tensor _segment_reduce_cpu_kernel(
...
+              counter++;
+            }
+            k++;
+          }
+          // Average gradient output based on number of maximum in the segment
+          TORCH_CHECK(counter > 0);
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:99]: This should be TORCH_INTERNAL_ASSERT

File: test/test_segment_reductions.py [hunk 5]
```diff
@@ -11,28 +11,40 @@
...
+        # Backward is only supported for cpu tensors for now. Return early if cuda
+        if data.is_cuda:
+            return
+
+        # Test backward
+        expected_grad = torch.tensor([1, 1, 0, 0, 0.5, 0.5], device=device, dtype=dtype)
+        actual_result.sum().backward()
         self.assertEqual(
-            expected_result, actual_result, rtol=1e-03, atol=1e-05, equal_nan=True
+            expected_grad, data.grad, rtol=1e-03, atol=1e-05, equal_nan=True
```
Comment (ngimel) [inline test/test_segment_reductions.py:38]: direct checking of gradients is good, but you should also test with `gradcheck`

Comment (serhaty) [reply test/test_segment_reductions.py:38]: is there an example I can check for this that works with ops that have key value arguments?

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 2 above)
Comment (serhaty) [reply aten/src/ATen/native/SegmentReduce.cpp:69]: Sounds good, updated.

File: test/test_segment_reductions.py [hunk 6]
```diff
@@ -7,32 +7,64 @@
...
         )
 
+        # gradcheck does not work well with bfloat16 or fp16 cpu types
+        # also there is small numerical difference with fp32
+        if dtype not in [torch.half, torch.bfloat16, torch.float]:
```
Comment (ngimel) [inline test/test_segment_reductions.py:44]: check grad only for double (`if dtype == torch.double`)

**Review by ngimel — APPROVED** [review]
Looks good, please consider making it structured in the followup, as @ezyang suggests

**Comment by serhaty** [comment]
This PS has been checked-in. Closing it manually.
